pkg/controllers/updaterun/validation.go

```diff
@@ -0,0 +1,319 @@
+			klog.ErrorS(unexpectedErr, "There's not yet started stage before the finished stage", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		// Record the last finished stage so we can continue from the next stage if no stage is updating.
+		lastFinishedStageIndex = curStage
+	} else if condition.IsConditionStatusFalse(stageSucceedCond, updateRun.Generation) {
+		// The stage has failed.
+		failedErr := fmt.Errorf("the stage `%s` has failed, err: %s", stageStatus.StageName, stageSucceedCond.Message)
+		klog.ErrorS(failedErr, "The stage has failed", "stageCond", stageSucceedCond, "clusterStagedUpdateRun", klog.KObj(updateRun))
+		return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, failedErr.Error())
+	} else if stageStartedCond != nil {
+		// The stage is still updating.
+		if updatingStageIndex != -1 {
+			// There should be only one stage updating at a time.
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the stage `%s` is updating, but there is already a stage with index `%d` updating", stageStatus.StageName, updatingStageIndex))
+			klog.ErrorS(unexpectedErr, "Detected more than one updating stages", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		if curStage != lastFinishedStageIndex+1 {
+			// The current updating stage is not right after the last finished stage.
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the updating stage `%s` is not right after the last finished stage with index `%d`", stageStatus.StageName, lastFinishedStageIndex))
+			klog.ErrorS(unexpectedErr, "There's not yet started stage before the updating stage", "clusterStagedUpdateRun", klog.KObj(updateRun))
+			return -1, -1, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		updatingStageIndex = curStage
+		// Collect the updating clusters.
+		var updatingClusters []string
+		for j := range stageStatus.Clusters {
+			clusterStartedCond := meta.FindStatusCondition(stageStatus.Clusters[j].Conditions, string(placementv1alpha1.ClusterUpdatingConditionStarted))
+			clusterFinishedCond := meta.FindStatusCondition(stageStatus.Clusters[j].Conditions, string(placementv1alpha1.ClusterUpdatingConditionSucceeded))
+			if condition.IsConditionStatusTrue(clusterStartedCond, updateRun.Generation) &&
+				!(condition.IsConditionStatusTrue(clusterFinishedCond, updateRun.Generation) || condition.IsConditionStatusFalse(clusterFinishedCond, updateRun.Generation)) {
+				updatingClusters = append(updatingClusters, stageStatus.Clusters[j].ClusterName)
+			}
+		}
+		// We don't allow more than one clusters to be updating at the same time.
```

Comment:
can u add a TODO to support more than one cluster updating at the same time.

Reply:
Added, thanks!